### DISCUSSION TITLE
support for multiple devices

### DISCUSSION THREAD

**Discussion by CharlesGodwin** [CharlesGodwin/pymagnum#25] [Ideas]
I am in the process of finalizing Release 2.0 and the topic of support for multiple serial devices (RS485) has come up. This is needed when using a multi-inverter configuration as each inverter needs its own RS485 device.

I propose to add support for multiple devices to the **magdump** dump program shipped with the package.

The solution I intend to implement is to allow the `-d `or` --device `option to be defined multiple times. You would use code like this:
`magdump --device /dev/ttyUSB0 --device /dev/ttyUSB1`
The program would iterate through the list of devices to produce output. The program currently supports a one time dump or can loop every  `--interval` seconds and this will continue but all devices will be queried at each iteration.

In order to allow downstream programs to distinguish data I am adding `comm_device: "/dev/USB0"` to the generate JSON for each device object.

 I have not committed to this yet and welcome input. I will also update the example programs to support the same implementation.

**Comment by Malkierian** [comment]
That seems a reasonable option. As long as someone makes a quick script to run in a terminal, it should be pretty easy to handle quick startups, too, and not have to remember devices or rely on terminal's memory. You'll be adding this to magserver as well, correct?

You said it would be complicated to add config stuff to it, but I don't even want to think about how messy config would be for an Influx integration, considering you wouldn't want to be feeding every bit of info to Influx that you can read from the Magnums...

**Reply by CharlesGodwin** [reply]
@Malkierian I've posted a [discussion ](https://github.com/CharlesGodwin/pymagnum/discussions/27)about a config file.

**Comment by CharlesGodwin** [comment]
I plan to upgrade the three example programs to match. The example programs are just that, examples. I provided them as stepping stones for developers to do their own work.

I use InfluxDB but all my code is specialized but built from the examples base. I have data sources from Magnum, a charge controller and a few "thermometer" sensors.
From Magnum all I pass to InfluxDB are data points like voltage, current. All the static stuff I ignore. I do analyse state changes, like generator start and stop, but log them seperatetly as events, rather than time series data. Most of the Magnum "data" is ignored. 

I'll try to start my configuration discussion later today.

**Comment by Malkierian** [comment]
I'm assuming multiple devices being specified adds them to something like an array of them in order of specification? If you don't specify options, do they default to those set as defaults in the program, or does it use other devices' specifications?

**Reply by CharlesGodwin** [reply]
Honestly, I never thought of discrete options for each device. These would be timeout and packets. The current implementation uses the same values for all devices. I'd like to see some testing results before I go down the rabbit hole of separate values for each. I'm not sure if they're needed.

I can, if asked, publish a beta version tomorrow. The only major shortcoming now is the documentation is not finished. And I have not settled the config file issue.
The coding for multiple devices is done, until bug reports prove me wrong.

**Comment by CharlesGodwin** [comment]
@Malkierian I have [published ](https://github.com/CharlesGodwin/pymagnum/discussions/30)a release candidate.

**Comment by kkrankall** [comment]
Charles,

I continue to use pymagnum and it is working great. I just got another 4024PAE to stack and was reading this discussion. Is there a way to read info from the 2nd inverter (Slave1)?

Do I need a 2nd pi or does the device name change in the json data from INVERTER to something else?

Thank you again for this great tool.

Kevin

**Reply by CharlesGodwin** [reply]
A second Pi is an easy, but not cheap, solution. You could also just go with a second RS485/USB or RS485 HAT. Of course getting a new Pi these days is a challenge.

Because of the way Magnum has things setup, I haven't found a way to get all the Inverter data on one RS485 connection.

Unless you write your own code using my examples, you will need to run two copies of the monitor programs, with separate device names for each. Like /dev/ttyUSB0 and /dev/ttyUSB1.

**EDIT** If you write your own, you just need two instances of the Magnum class, one for each device. The read both each cycle.

Write if you need more help

**Reply by kkrankall** [reply]
Awesome. I will give that a try once I install the 2nd inverter.

**Reply by Malkierian** [reply]
Yes, what Charles mentioned is what I did. I went in the code and set up multiple simultaneous Magnum device tracking. It was rather simple, if you know programming.

**Reply by kkrankall** [reply]
Working great. Was easy to setup the additional RS485 usb dongle.

 I have everything setup in HA. Thank you all

<img width="787" alt="Screen Shot 2022-10-27 at 9 25 31 AM" src="https://user-images.githubusercontent.com/15572661/198297211-24497338-aa3c-47dd-a304-5b265d3a098b.png">

**Reply by CharlesGodwin** [reply]
@kkrankall  I neglected to add in my initial reply that starting in version 2.0 `magdump` and the example programs support defining multiple `--device` parameters and the code will handle the iteration. [Details here](https://pymagnum.readthedocs.io/en/latest/tools.html?highlight=multiple#magdump:~:text=device.%20Just%20provide-,multiple,-%2D%2Ddevice%20/dev/ttyUSBn)

**Reply by kkrankall** [reply]
Yes I picked up on that. It's good that you updated the documentation. 

Thanks again it's working great.
